отправляю на ревью задачу BJS2-3413

semyonsss opened the pull request.

semyonsss committed:
отправляю на ревью задачу BJS2-3413

bukevsky commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,33 @@
+public class VacancyController {
+    private VacancyService vacancyService;
+
+    public VacancyDto createVacancy(VacancyDto vacancyDto) {
+        return vacancyService.createVacancy(vacancyDto);
+    }
+
+    public VacancyDto updateVacancy(VacancyDto vacancyDto) {
+        return vacancyService.updateVacancy(vacancyDto);
+    }
+
+    public VacancyDto deleteVacancy(VacancyDto vacancyDto) {
+        return vacancyService.deleteVacancy(vacancyDto);
+    }
+
+    public List<VacancyDto> getFiltersVacancies(VacancyFilterDto vacancyFilterDto) {
+        return vacancyService.getFilterVacancies(vacancyFilterDto);
+    }
+
+    public VacancyDto getVacancy(Long vacancyId) {
+        return vacancyService.getVacancy(vacancyId);
+    }
```
Мы сейчас прошли такие аннотации как @GetMapping, @PostMapping, @PutMapping, @DeleteMapping,
Добавь их где они нужны. Так же можешь над классом повесить @RequestMapping("/vacancy")

bukevsky replied:
Так же можешь тестить свои методы Postman'ом, на самом деле, думаешь все работает правильно, а может оказаться метод репозитория например неправильно работает и не делает свои функции. Советую поскорее попробовать им попользоваться!

semyonsss replied:
Да спасибо посомотрел лекцию по аннотациям, раствил их

bukevsky commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,82 @@
+
+    public VacancyDto updateVacancy(VacancyDto vacancyDto) {
+        Vacancy vacancy = vacancyRepository.findById(vacancyDto.getId()).orElseThrow(() -> new EntityNotFoundException("Вакансия не найдена"));
+        if (vacancy.getStatus() == VacancyStatus.CLOSED) {
```
Сравнение константы слева должно быть 
VacancyStatus.CLOSED.equals(vacancy.getStatus())
почитай про это правило: "Сравнение константы слева"

semyonsss replied:
ммм Спасибо буду знать, очень полезно

bukevsky commented on src/main/java/faang/school/projectservice/validation/VacancyValidation.java:
```diff
@@ -0,0 +1,77 @@
+        Vacancy vacancy = vacancyRepository.findById(vacancyDto.getId()).orElseThrow(() -> new EntityNotFoundException("Вакансия не найдена"));
+        int totalVacancies = vacancy.getCount();
+        long candidatesApprove = vacancy.getCandidates().stream()
+                .filter(candidate -> CandidateStatus.ACCEPTED.equals(candidate.getCandidateStatus()))
```
а тут правильно, хорош

bukevsky commented on src/main/java/faang/school/projectservice/validation/VacancyValidation.java:
```diff
@@ -0,0 +1,77 @@
+
+        List<Long> candidates = vacancy.getCandidates().stream().map(candidate -> candidate.getUserId()).toList();
+        vacancy.getProject().getTeams().stream()
+                .flatMap(team -> team.getTeamMembers().stream())
+                .filter(member -> candidates.contains(member.getId()))
+                .forEach(teamMember -> {
+                    if (teamMember.getRoles() == null || teamMember.getRoles().isEmpty()) {
+                        throw new DataVacancyValidation("Вакансия не может быть закрыта, пока вся команда не получит свои роли");
+                    }
+                });
+    }
+}
```
Все месседжи ошибок можно вынести в отдельный enum class

semyonsss replied:
Кул, сделал так спасибо)

bukevsky reviewed (commented):
Хорошая работа!

semyonsss committed:
внес правки
добавил аннотации методов запроса в контроллере, создал enum для сообщений для ошибок

Vahe650 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public VacancyDto deleteVacancy(VacancyDto vacancyDto) {
+        Vacancy vacancy = vacancyRepository.findById(vacancyDto.getId()).orElseThrow(() -> new EntityNotFoundException("Вакансия не найдена"));
```
для болшей информации добавь идшник в сообшение
new EntityNotFoundException("Вакансия не найдена"+ vacancyDto.getId())

semyonsss replied:
Хорошо, спасибо

Vahe650 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,82 @@
+
+    public VacancyDto getVacancy(Long vacancyId) {
+        Vacancy vacancy = vacancyRepository.findById(vacancyId).orElseThrow(() -> new EntityNotFoundException("Вакансия не найдена"));
+        return vacancyMapper.toDto(vacancy);
```
добавь идшник в сообшение

Vahe650 commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,82 @@
+        return vacancyDtoList;
+    }
+
+    public VacancyDto getVacancy(Long vacancyId) {
```
название метода должно отражать само действие метода
getVacancyById

Vahe650 commented on src/main/java/faang/school/projectservice/validation/VacancyValidation.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public void validationVacancyClosed(VacancyDto vacancyDto) {
+        Vacancy vacancy = vacancyRepository.findById(vacancyDto.getId()).orElseThrow(() -> new EntityNotFoundException("Вакансия не найдена"));
```
добавь идшник в сообшение

Vahe650 commented on src/test/java/faang/school/projectservice/filter/VacancyNameFilterTest.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Test
+    public void testIsApplicable() {
```
напиши eще тест 
    public void testIsNotApplicable() {

Vahe650 commented on src/test/java/faang/school/projectservice/service/VacancyServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+
+
+    @Test
+    public void testCreateVacancySave() {
```
назови метод testCreateVacancySaveSuccess и напиши еще тест метод testCreateVacancySaveFailure

semyonsss committed:
Внес изменения после ревью
